File: src/PlatynUI/ui/strategies/element.py
# SPDX-FileCopyrightText: 2024 Daniel Biehl <daniel.biehl@imbus.de>
#
# SPDX-License-Identifier: Apache-2.0

from typing import ClassVar, Optional

from ...core import Adapter, Point, Rect, StrategyBase

__all__ = ["Element", "TopLevelElement"]


class Element(StrategyBase):
    strategy_name: ClassVar[str] = "org.platynui.strategies.Element"

    @property
    def is_readonly(self) -> bool:
        raise NotImplementedError

    @property
    def is_enabled(self) -> bool:
        raise NotImplementedError

    @property
    def is_visible(self) -> bool:
        raise NotImplementedError

    @property
    def is_in_view(self) -> bool:
        raise NotImplementedError

    @property
    def toplevel_parent_is_active(self) -> bool:
        raise NotImplementedError

    @property
    def bounding_rectangle(self) -> Rect:
        raise NotImplementedError

    @property
    def visible_rectangle(self) -> Rect:
        raise NotImplementedError

    @property
    def top_level_parent(self) -> Optional[Adapter]:
        raise NotImplementedError

    @property
    def default_click_position(self) -> Point:
        raise NotImplementedError

    def try_ensure_visible(self) -> bool:
        raise NotImplementedError

    def try_ensure_application_is_ready(self) -> bool:
        raise NotImplementedError

    def try_ensure_toplevel_parent_is_active(self) -> bool:
        raise NotImplementedError

    def try_bring_into_view(self) -> bool:
        raise NotImplementedError


class TopLevelElement(StrategyBase):
    strategy_name = "org.platynui.strategies.TopLevelElement"

    @property
    def is_top_level_element(self) -> bool:
        raise NotImplementedError

File: src/PlatynUI.Runtime/Core/Element.cs
// SPDX-FileCopyrightText: 2024 Daniel Biehl <daniel.biehl@imbus.de>
//
// SPDX-License-Identifier: Apache-2.0

namespace PlatynUI.Runtime.Core;

public interface IElement
{
    bool IsEnabled { get; }
    bool IsVisible { get; }
    bool IsInView { get; }
    bool TopLevelParentIsActive { get; }
    Rect BoundingRectangle { get; }
    Rect VisibleRectangle { get; }
    Point? DefaultClickPosition { get; }

    bool TryEnsureVisible();
    bool TryEnsureApplicationIsReady();
    bool TryEnsureToplevelParentIsActive();
    bool TryBringIntoView();
}
